### ISSUE TITLE
0.1 + 0.2 = 0.30000000000000004

### ISSUE THREAD

**Issue by realeatham** [sosauce/Vanilla#66]
**Describe the bug**
See title 

**To Reproduce**
See title 

**Expected behavior**
The sum to equal 0.3

**Screenshots**

**Smartphone (please complete the following information):**
 - Device: s21
 - OS: android 14 I think 
 -app Version: 2.7.1

**Additional context**
Could also be a problem with keval @notKamui

**Comment by sosauce** [comment]
https://discuss.kotlinlang.org/t/kotlin-compiler-apparently-losing-double-value-precision/14076

See this, I think it's "normal" and nothing we can do about it

**Comment by sosauce** [comment]
https://floating-point-gui.de/

**Comment by sosauce** [comment]
Oh wait, I could make it refactor or sum but do I really want to?

**Comment by notKamui** [comment]
@realeatham this is completely normal behavior for floating point arithmetic in binary representation. Please try that in many other languages and you'll see the same result.

**Comment by SlamaFR** [comment]
@realeatham So this is basically a well known issue of floating point numbers in computer science.

See https://en.wikipedia.org/wiki/IEEE_754#Rounding_rules

**Comment by hewantstocomment** [comment]
<img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/3c084f24-4c70-458f-8d18-dca76561f36e" />
<img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/8b6cd87a-3f94-434d-9ac4-f999f7a25fed" />

I guess it's the same problem...
Expexcted result of 8+12+8+22.12 is 50.12  ...
Device: Pixel 8a on GrapheneOS. Everything up-to-date.
Maybe add a warning to the app?
e.g. in the description of the app-stores and here write that: there is some limitation when it comes to decimal numbers usw.


Anyway, thanks for the app and your work!

**Comment by theoholl** [comment]
> https://discuss.kotlinlang.org/t/kotlin-compiler-apparently-losing-double-value-precision/14076
> 
> See this, I think it's "normal" and nothing we can do about it

Hey, thank you for your open source contributions! I think we should do something about this issue. Because from a user's perspective, it may be really confusing facing all these unexpected zeroes and could make the app seem broken. Do you agree?

Now that we understand the technical reasons for the problem, we can think about how to solve it. My initial thoughts are:

1. Rounding: we could define the required/supported precision and apply a [rounding strategy](https://floating-point-gui.de/errors/rounding/). With rounding, we may lose some precision, e.g. `0.00000000000000001 + 0.00000000000000003 = 0`. I'm not a mathematician but I guess we have a similar challenge when handling irrational numbers like pi.
2. Extend [Keval](https://github.com/notKamui/Keval) to support [BigDecimal](https://developer.android.com/reference/kotlin/android/icu/math/BigDecimal). This may not be as trivial as adding `BigDecimal` method overloads because Keval uses [`kotlin.math`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.math/) for calculations that operates on `Double`/`Float`. So maybe we'd have to add something like [`big-math`](https://github.com/eobermuhlner/big-math).
3. Switch out Keval for another library like [EvalEx](https://github.com/ezylang/EvalEx)

Do you see a different approach? :) 

Rounding might be a practical approach. I really like Keval for its simplicity but using a more actively maintained library might be a good idea, too. @notKamui, would you be open to review PRs?

**Comment by notKamui** [comment]
@theoholl  yes I am open to PRs. the main issue is publishing though. Maven Central and Gradle have been a pain in the ass since the last publish and I struggle to find the time to fix that.

any help is appreciated though

**Comment by sosauce** [comment]
Hey @theoholl Kamui already said he'd be working on a rewrite to use BigDecimal, i don't have a real reason to switch away from Keval I like the library and Kamui has helped me multiple times when I needed

**Comment by notKamui** [comment]
@sosauce 2.0.0 is out with support for generic output via adapter implementation. Watch out for breaking changes. Don't hesitate to ask for help or if there is a bug. Do re-read the README.md

**Comment by sosauce** [comment]
Check v4.1.0 should be all good now
